File: experiment/harness/plan.py
from __future__ import annotations

import argparse
from itertools import permutations
import random
from typing import Any

from .common import load_json, opaque_id, resolve_run, utc_now, write_json

ALL_ROLES = [
    "simplicity",
    "correctness",
    "architecture",
    "security",
    "cost",
    "entropy",
    "language",
]


class Builder:
    def __init__(self, config: dict[str, Any]):
        self.config = config
        self.seed = config["seed"]
        self.calls: list[dict[str, Any]] = []
        self.sets: list[dict[str, Any]] = []

    def call(
        self,
        semantic: tuple[object, ...],
        *,
        packet: str,
        provider: dict[str, Any],
        prompt_spec: dict[str, Any],
        depends_on: list[str] | None = None,
        phase: str = "review",
    ) -> dict[str, Any]:
        call_id = "c-" + opaque_id(self.seed, "call", *semantic)
        value = {
            "call_id": call_id,
            "reviewer_id": "r-" + opaque_id(self.seed, "reviewer", *semantic, length=10),
            "packet": packet,
            "provider": provider,
            "prompt_spec": prompt_spec,
            "depends_on": depends_on or [],
            "phase": phase,
            "semantic": list(semantic),
        }
        self.calls.append(value)
        return value

    def output_set(
        self,
        semantic: tuple[object, ...],
        *,
        packet: str,
        call_ids: list[str],
        cost_call_ids: list[str],
        metadata: dict[str, Any],
    ) -> None:
        self.sets.append(
            {
                "set_id": "s-" + opaque_id(self.seed, "set", *semantic),
                "packet": packet,
                "call_ids": call_ids,
                "cost_call_ids": cost_call_ids,
                "metadata": metadata,
                "semantic": list(semantic),
            }
        )


def stage_a(builder: Builder) -> None:
    config = builder.config
    provider = config["provider"]
    for packet in config["packets"]:
        for repeat in range(config["repetitions"]):
            for arm in config["arms"]:
                prefix = ("stage_a", packet, repeat, arm)
                if arm in {"S0", "S1", "S2"}:
                    spec = {
                        "S0": {"kind": "generic"},
                        "S1": {"kind": "omnibus", "wrapper": "functional"},
                        "S2": {"kind": "omnibus", "wrapper": "fictional"},
                    }[arm]
                    review = builder.call(prefix + ("review",), packet=packet, provider=provider, prompt_spec=spec)
                    builder.output_set(
                        prefix + ("final",),
                        packet=packet,
                        call_ids=[review["call_id"]],
                        cost_call_ids=[review["call_id"]],
                        metadata={"design": "stage_a", "arm": arm, "repeat": repeat, "kind": "final"},
                    )
                    continue

                reviews: list[dict[str, Any]] = []
                for index, role in enumerate(ALL_ROLES):
                    if arm == "M0":
                        spec = {"kind": "omnibus", "wrapper": "functional", "sample": index}
                    elif arm == "M1":
                        spec = {"kind": "specialist", "wrapper": "functional", "role": role}
                    elif arm == "M2":
                        spec = {"kind": "specialist", "wrapper": "fictional", "role": role}
                    else:
                        raise ValueError(f"Unknown Stage A arm: {arm}")
                    reviews.append(
                        builder.call(prefix + ("review", index, role), packet=packet, provider=provider, prompt_spec=spec)
                    )
                review_ids = [review["call_id"] for review in reviews]
                fused = builder.call(
                    prefix + ("fuse",),
                    packet=packet,
                    provider=provider,
                    prompt_spec={"kind": "fuser"},
                    depends_on=review_ids,
                    phase="fuse",
                )
                builder.output_set(
                    prefix + ("raw_union",),
                    packet=packet,
                    call_ids=review_ids,
                    cost_call_ids=review_ids,
                    metadata={"design": "stage_a", "arm": arm, "repeat": repeat, "kind": "raw_union"},
                )
                builder.output_set(
                    prefix + ("fused",),
                    packet=packet,
                    call_ids=[fused["call_id"]],
                    cost_call_ids=review_ids + [fused["call_id"]],
                    metadata={"design": "stage_a", "arm": arm, "repeat": repeat, "kind": "fused"},
                )


def topology(builder: Builder) -> None:
    config = builder.config
    provider = config["provider"]
    roles = config["roles"]
    orders = list(permutations(roles)) if config.get("all_role_orders") else [tuple(roles)]
    wrapper = config["wrapper"]
    for packet in config["packets"]:
        for repeat in range(config["repetitions"]):
            for order in orders:
                order_name = "-".join(order)
                for topology_name in config["topologies"]:
                    prefix = ("topology", packet, repeat, order_name, topology_name)
                    reviews: list[dict[str, Any]] = []
                    previous: list[str] = []
                    for hop, role in enumerate(order):
                        dynamic = topology_name == "chain" and hop > 0
                        spec = {
                            "kind": "chain" if dynamic else "specialist",
                            "wrapper": wrapper,
                            "role": role,
                            "hop": hop + 1,
                        }
                        dependencies = previous[-1:] if dynamic else []
                        review = builder.call(
                            prefix + ("review", hop, role),
                            packet=packet,
                            provider=provider,
                            prompt_spec=spec,
                            depends_on=dependencies,
                        )
                        reviews.append(review)
                        previous.append(review["call_id"])
                        if topology_name == "chain":
                            builder.output_set(
                                prefix + (f"hop-{hop + 1}",),
                                packet=packet,
                                call_ids=[review["call_id"]],
                                cost_call_ids=previous.copy(),
                                metadata={
                                    "design": "topology",
                                    "topology": topology_name,
                                    "order": order_name,
                                    "repeat": repeat,
                                    "kind": f"hop-{hop + 1}",
                                },
                            )
                    review_ids = [review["call_id"] for review in reviews]
                    fused = builder.call(
                        prefix + ("fuse",),
                        packet=packet,
                        provider=provider,
                        prompt_spec={"kind": "fuser"},
                        depends_on=review_ids,
                        phase="fuse",
                    )
                    if topology_name == "fanout":
                        builder.output_set(
                            prefix + ("raw_union",),
                            packet=packet,
                            call_ids=review_ids,
                            cost_call_ids=review_ids,
                            metadata={
                                "design": "topology",
                                "topology": topology_name,
                                "order": order_name,
                                "repeat": repeat,
                                "kind": "raw_union",
                            },
                        )
                    builder.output_set(
                        prefix + ("fused",),
                        packet=packet,
                        call_ids=[fused["call_id"]],
                        cost_call_ids=review_ids + [fused["call_id"]],
                        metadata={
                            "design": "topology",
                            "topology": topology_name,
                            "order": order_name,
                            "repeat": repeat,
                            "kind": "fused",
                        },
                    )


def provider_pair(builder: Builder) -> None:
    config = builder.config
    for packet in config["packets"]:
        for repeat in range(config["repetitions"]):
            for provider_index, provider in enumerate(config["providers"]):
                for wrapper in config["wrappers"]:
                    prefix = ("provider_pair", packet, repeat, provider_index, wrapper)
                    review = builder.call(
                        prefix + ("review",),
                        packet=packet,
                        provider=provider,
                        prompt_spec={"kind": "specialist", "wrapper": wrapper, "role": config["role"]},
                    )
                    builder.output_set(
                        prefix + ("final",),
                        packet=packet,
                        call_ids=[review["call_id"]],
                        cost_call_ids=[review["call_id"]],
                        metadata={
                            "design": "provider_pair",
                            "provider_index": provider_index,
                            "adapter": provider["adapter"],
                            "model": provider["model"],
                            "wrapper": wrapper,
                            "repeat": repeat,
                            "kind": "final",
                        },
                    )


def build(config: dict[str, Any]) -> dict[str, Any]:
    builder = Builder(config)
    design = config["design"]
    if design == "stage_a":
        stage_a(builder)
    elif design == "topology":
        topology(builder)
    elif design == "provider_pair":
        provider_pair(builder)
    else:
        raise ValueError(f"Unknown design: {design}")
    random.Random(config["seed"]).shuffle(builder.calls)
    return {
        "schema_version": 1,
        "created_at": utc_now(),
        "design": design,
        "seed": config["seed"],
        "calls": builder.calls,
        "output_sets": builder.sets,
        "counts": {"calls": len(builder.calls), "output_sets": len(builder.sets)},
    }


def main() -> int:
    parser = argparse.ArgumentParser(description="Create a deterministic call plan for a frozen run")
    parser.add_argument("run")
    args = parser.parse_args()
    run = resolve_run(args.run)
    if (run / "plan.json").exists():
        raise SystemExit("plan.json already exists; do not replace a frozen run plan")
    config = load_json(run / "config.json")
    plan = build(config)
    write_json(run / "plan.json", plan)
    print(f"Planned {plan['counts']['calls']} calls and {plan['counts']['output_sets']} output sets.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: experiment/harness/plan.go
package harness

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

var allRoles = []string{"simplicity", "correctness", "architecture", "security", "cost", "entropy", "language"}

type builder struct {
	config Config
	calls  []Call
	sets   []OutputSet
}

func semanticParts(prefix []any, extra ...any) []any {
	out := append([]any{}, prefix...)
	return append(out, extra...)
}
func (b *builder) call(semantic []any, packet string, provider Provider, spec map[string]any, depends []string, phase string) Call {
	if phase == "" {
		phase = "review"
	}
	idParts := append([]any{"call"}, semantic...)
	reviewerParts := append([]any{"reviewer"}, semantic...)
	call := Call{CallID: "c-" + opaqueID(b.config.Seed, 16, idParts...), ReviewerID: "r-" + opaqueID(b.config.Seed, 10, reviewerParts...), Packet: packet, Provider: provider, PromptSpec: spec, DependsOn: append([]string{}, depends...), Phase: phase, Semantic: semantic}
	b.calls = append(b.calls, call)
	return call
}
func (b *builder) outputSet(semantic []any, packet string, callIDs, costIDs []string, metadata map[string]any) {
	parts := append([]any{"set"}, semantic...)
	b.sets = append(b.sets, OutputSet{SetID: "s-" + opaqueID(b.config.Seed, 16, parts...), Packet: packet, CallIDs: append([]string{}, callIDs...), CostCallIDs: append([]string{}, costIDs...), Metadata: metadata, Semantic: semantic})
}

func buildStageA(b *builder) error {
	c := b.config
	for _, packet := range c.Packets {
		for repeat := 0; repeat < c.Repetitions; repeat++ {
			for _, arm := range c.Arms {
				prefix := []any{"stage_a", packet, repeat, arm}
				if arm == "S0" || arm == "S1" || arm == "S2" {
					var spec map[string]any
					if arm == "S0" {
						spec = map[string]any{"kind": "generic"}
					} else if arm == "S1" {
						spec = map[string]any{"kind": "omnibus", "wrapper": "functional"}
					} else {
						spec = map[string]any{"kind": "omnibus", "wrapper": "fictional"}
					}
					review := b.call(semanticParts(prefix, "review"), packet, c.Provider, spec, nil, "")
					b.outputSet(semanticParts(prefix, "final"), packet, []string{review.CallID}, []string{review.CallID}, map[string]any{"design": "stage_a", "arm": arm, "repeat": repeat, "kind": "final"})
					continue
				}
				reviews := []Call{}
				for index, role := range allRoles {
					var spec map[string]any
					switch arm {
					case "M0":
						spec = map[string]any{"kind": "omnibus", "wrapper": "functional", "sample": index}
					case "M1":
						spec = map[string]any{"kind": "specialist", "wrapper": "functional", "role": role}
					case "M2":
						spec = map[string]any{"kind": "specialist", "wrapper": "fictional", "role": role}
					default:
						return fmt.Errorf("unknown Stage A arm %s", arm)
					}
					reviews = append(reviews, b.call(semanticParts(prefix, "review", index, role), packet, c.Provider, spec, nil, ""))
				}
				ids := callIDs(reviews)
				fused := b.call(semanticParts(prefix, "fuse"), packet, c.Provider, map[string]any{"kind": "fuser"}, ids, "fuse")
				b.outputSet(semanticParts(prefix, "raw_union"), packet, ids, ids, map[string]any{"design": "stage_a", "arm": arm, "repeat": repeat, "kind": "raw_union"})
				b.outputSet(semanticParts(prefix, "fused"), packet, []string{fused.CallID}, append(append([]string{}, ids...), fused.CallID), map[string]any{"design": "stage_a", "arm": arm, "repeat": repeat, "kind": "fused"})
			}
		}
	}
	return nil
}
func callIDs(calls []Call) []string {
	out := make([]string, len(calls))
	for i, c := range calls {
		out[i] = c.CallID
	}
	return out
}

func permutations(values []string) [][]string {
	if len(values) == 0 {
		return [][]string{{}}
	}
	out := [][]string{}
	for i, v := range values {
		rest := append([]string{}, values[:i]...)
		rest = append(rest, values[i+1:]...)
		for _, tail := range permutations(rest) {
			out = append(out, append([]string{v}, tail...))
		}
	}
	return out
}
func buildTopology(b *builder) error {
	c := b.config
	orders := [][]string{append([]string{}, c.Roles...)}
	if c.AllRoleOrders {
		orders = permutations(c.Roles)
	}
	for _, packet := range c.Packets {
		for repeat := 0; repeat < c.Repetitions; repeat++ {
			for _, order := range orders {
				orderName := strings.Join(order, "-")
				for _, topology := range c.Topologies {
					prefix := []any{"topology", packet, repeat, orderName, topology}
					reviews := []Call{}
					previous := []string{}
					for hop, role := range order {
						dynamic := topology == "chain" && hop > 0
						kind := "specialist"
						if dynamic {
							kind = "chain"
						}
						depends := []string{}
						if dynamic {
							depends = []string{previous[len(previous)-1]}
						}
						review := b.call(semanticParts(prefix, "review", hop, role), packet, c.Provider, map[string]any{"kind": kind, "wrapper": c.Wrapper, "role": role, "hop": hop + 1}, depends, "")
						reviews = append(reviews, review)
						previous = append(previous, review.CallID)
						if topology == "chain" {
							b.outputSet(semanticParts(prefix, fmt.Sprintf("hop-%d", hop+1)), packet, []string{review.CallID}, append([]string{}, previous...), map[string]any{"design": "topology", "topology": topology, "order": orderName, "repeat": repeat, "kind": fmt.Sprintf("hop-%d", hop+1)})
						}
					}
					ids := callIDs(reviews)
					fused := b.call(semanticParts(prefix, "fuse"), packet, c.Provider, map[string]any{"kind": "fuser"}, ids, "fuse")
					if topology == "fanout" {
						b.outputSet(semanticParts(prefix, "raw_union"), packet, ids, ids, map[string]any{"design": "topology", "topology": topology, "order": orderName, "repeat": repeat, "kind": "raw_union"})
					}
					b.outputSet(semanticParts(prefix, "fused"), packet, []string{fused.CallID}, append(append([]string{}, ids...), fused.CallID), map[string]any{"design": "topology", "topology": topology, "order": orderName, "repeat": repeat, "kind": "fused"})
				}
			}
		}
	}
	return nil
}
func buildProviderPair(b *builder) error {
	c := b.config
	for _, packet := range c.Packets {
		for repeat := 0; repeat < c.Repetitions; repeat++ {
			for providerIndex, provider := range c.Providers {
				for _, wrapper := range c.Wrappers {
					prefix := []any{"provider_pair", packet, repeat, providerIndex, wrapper}
					review := b.call(semanticParts(prefix, "review"), packet, provider, map[string]any{"kind": "specialist", "wrapper": wrapper, "role": c.Role}, nil, "")
					b.outputSet(semanticParts(prefix, "final"), packet, []string{review.CallID}, []string{review.CallID}, map[string]any{"design": "provider_pair", "provider_index": providerIndex, "adapter": provider.Adapter, "model": provider.Model, "wrapper": wrapper, "role": c.Role, "repeat": repeat, "kind": "final"})
				}
			}
		}
	}
	return nil
}

func buildPersonaFactorial(b *builder) error {
	c := b.config
	for _, packet := range c.Packets {
		for repeat := 0; repeat < c.Repetitions; repeat++ {
			for _, role := range c.Roles {
				for _, wrapper := range c.Wrappers {
					prefix := []any{"persona_factorial", packet, repeat, role, wrapper}
					spec := map[string]any{"kind": "specialist", "wrapper": wrapper, "role": role}
					if role == "omnibus" {
						spec = map[string]any{"kind": "omnibus", "wrapper": wrapper}
					}
					review := b.call(semanticParts(prefix, "review"), packet, c.Provider, spec, nil, "")
					b.outputSet(semanticParts(prefix, "final"), packet, []string{review.CallID}, []string{review.CallID}, map[string]any{
						"design": "persona_factorial", "provider_index": 0, "adapter": c.Provider.Adapter,
						"model": c.Provider.Model, "wrapper": wrapper, "role": role, "repeat": repeat, "kind": "final",
					})
				}
			}
		}
	}
	return nil
}

func BuildPlan(config Config) (Plan, error) {
	b := &builder{config: config}
	var err error
	switch config.Design {
	case "stage_a":
		err = buildStageA(b)
	case "topology":
		err = buildTopology(b)
	case "provider_pair":
		err = buildProviderPair(b)
	case "persona_factorial":
		err = buildPersonaFactorial(b)
	default:
		err = fmt.Errorf("unknown design %s", config.Design)
	}
	if err != nil {
		return Plan{}, err
	}
	sort.SliceStable(b.calls, func(i, j int) bool {
		return shaText(config.Seed+"\x1forder\x1f"+b.calls[i].CallID) < shaText(config.Seed+"\x1forder\x1f"+b.calls[j].CallID)
	})
	return Plan{SchemaVersion: 2, CreatedAt: utcNow(), Design: config.Design, Seed: config.Seed, Calls: b.calls, OutputSets: b.sets, Counts: map[string]int{"calls": len(b.calls), "output_sets": len(b.sets)}}, nil
}
func (h *Harness) Plan(run string) error {
	resolved, err := h.resolveRun(run)
	if err != nil {
		return err
	}
	path := filepath.Join(resolved, "plan.json")
	if _, err := os.Stat(path); err == nil {
		return fmt.Errorf("plan.json already exists; do not replace a frozen run plan")
	}
	var config Config
	if err := readJSON(filepath.Join(resolved, "config.json"), &config); err != nil {
		return err
	}
	plan, err := BuildPlan(config)
	if err != nil {
		return err
	}
	if err := writeJSON(path, plan); err != nil {
		return err
	}
	fmt.Printf("Planned %d calls and %d output sets.\n", plan.Counts["calls"], plan.Counts["output_sets"])
	return nil
}
